fusequery/query/src/optimizers/optimizer_scatters.rs

```diff
@@ -0,0 +1,206 @@
+                    kind: StageKind::Convergent,
+                    scatters_expr: Expression::Literal(DataValue::UInt64(Some(0))),
+                    input: Arc::new(plan.clone())
+                }))
+            }
+        }
+    }
+
+    fn optimize_read_plan(
+        &mut self,
+        plan: &ReadDataSourcePlan,
+        status: &mut Vec<OptimizeKind>
+    ) -> Result<PlanNode> {
+        let read_table = self
+            .ctx
+            .get_datasource()
+            .get_table(plan.db.as_str(), plan.table.as_str())?;
+
+        let rows_threshold = self.ctx.get_min_distributed_rows()? as usize;
+        let bytes_threshold = self.ctx.get_min_distributed_bytes()? as usize;
+        if read_table.is_local()
+            && (plan.statistics.read_rows >= rows_threshold
+                || plan.statistics.read_bytes >= bytes_threshold)
+        {
+            // TODO: Need implement blockNumber function.
+            // We use the blockNumber function as the scatter expr.
+            // This will be DataBlock based round robin scheduling.
+            status.push(OptimizeKind::Scattered);
+            return Ok(PlanNode::Stage(StagePlan {
+                kind: StageKind::Expansive,
+                scatters_expr: Expression::ScalarFunction {
+                    op: String::from("blockNumber"),
+                    args: vec![]
+                },
+                input: Arc::new(PlanNode::ReadSource(plan.clone()))
+            }));
+        }
+
+        if read_table.is_local() {
+            status.push(OptimizeKind::Local);
+        } else {
+            status.push(OptimizeKind::Scattered);
+        }
+
+        // Keep local mode, there are two cases
+        // 1. The data is small enough that we just need to continue to run on standalone mode.
+        // 2. The table is already clustered. We just need to repartition for the cluster mode.
+        Ok(PlanNode::ReadSource(plan.clone()))
+    }
+
+    fn optimize_aggregator(
+        &mut self,
+        plan: &AggregatorPartialPlan,
+        input: PlanNode,
+        status: &mut Vec<OptimizeKind>
+    ) -> Result<PlanNode> {
+        if let Some(OptimizeKind::Local) = status.pop() {
+            // Keep running in standalone mode
+            // TODO We can estimate and evaluate the amount of data to decide whether to shuffle again
+            // For example, when ReadSourcePlan = 500MB
+            // We can evaluate 500MB * 0.8(it can be get by history) after FilterPlan
+            status.push(OptimizeKind::Local);
+            return Ok(PlanNode::AggregatorPartial(AggregatorPartialPlan {
+                group_expr: plan.group_expr.clone(),
+                aggr_expr: plan.aggr_expr.clone(),
+                schema: plan.schema.clone(),
+                input: Arc::new(input)
+            }));
+        }
+
+        match plan.group_expr.len() {
+            0 => {
+                // For the final state, we need to aggregate the data
```

Comment:
This comments made some confusion, does it mean:
If no `group by` we convergent it in local node?